File: ios/Classes/PapagaioTtsPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class PapagaioTtsPlugin: NSObject, FlutterPlugin {
  let tts = PapagaioTts()
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "papagaio_tts", binaryMessenger: registrar.messenger())
    let instance = PapagaioTtsPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getAvailableLanguages":
      // TODO
      let filter = call.arguments as! [String]
      let convertedFilter = filter.map { $0.replacingOccurrences(of: "_", with: "-", options: .literal, range: nil) }
      result(getAvailableLanguages(convertedFilter))
    case "getVoices":
      result(getVoices())
    case "speak":
      let text = call.arguments as! String
      speak(text)
    case "stop":
      stop()
    case "shutdown":
      stop()
    case "getSpeakingStatus":
      result(getSpeakingStatus())

    // Getters
    case "getLanguage":
      result(getLanguage())
    case "getVoice":
      result(getVoice())
    case "getRate":
      result(getRate())
    case "getVolume":
      result(getVolume())
    case "getPitch":
      result(getPitch())

    // Setters
    case "setVoice":
      let voice = call.arguments as! String
      setVoice(voice)
        result(true)
    case "setRate":
      let rate = (call.arguments as! NSNumber).floatValue
      setRate(rate)
        result(true)
    case "setLanguage":
      let language = call.arguments as! [String]
      setLanguage(language)
      result(true)
    case "setVolume":
      let volume = (call.arguments as! NSNumber).floatValue
      print("### pluginsetVolume \(volume)")
      setVolume(volume)
        result(true)
    case "setPitch":
      let pitch = (call.arguments as! NSNumber).floatValue
      setPitch(pitch)
        result(true)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  func getAvailableLanguages(_ filter: [String]) -> [[String]] {
    // TODO
    return tts.getAvailableLanguages(filter)
  }

  func getVoices() -> [String] {
      return tts.getVoices()
  }

  func speak(_ text: String) {
    return tts.speak(text)
  }

  func stop() {
    return tts.stop()
  }

  func getSpeakingStatus() -> Bool {
    return tts.getSpeakingStatus()
  }

  func getVoice() -> String {
    return tts.getVoice()
  }

  func getLanguage() -> String {
    return tts.getLanguage()
  }

  func getRate() -> Float {
    return tts.getRate()
  }

    func getVolume() -> Float {
    return tts.getVolume()
  }

  func getPitch() -> Float {
    return tts.getPitch()
  }


  func setVoice(_ voice: String) {
    return tts.setVoice(voice)
  }

  func setRate(_ rate: Float){
    return tts.setRate(rate)
  }

  func setLanguage(_ language: [String]) {
    if (language.count == 2) {
      return tts.setLanguage("\(language[0])-\(language[1])")
    }
    return tts.setLanguage("\(language[0]))")
  }

  func setVolume(_ volume: Float) {
    return tts.setVolume(volume)
  }

  func setPitch(_ pitch: Float) {
    return tts.setPitch(pitch)
  }
}

File: android/src/main/kotlin/papagaiojp/webnode/jp/papagaio_tts/PapagaioTtsPlugin.kt
package papagaiojp.webnode.jp.papagaio_tts

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.util.*

/** PapagaioTtsPlugin */
class PapagaioTtsPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var tts: PapagaioTts
    private var volume: Float = 1.0f
    private var rate: Float = 0.5f
    private var pitch: Float = 0.5f

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        tts = PapagaioTts(flutterPluginBinding.applicationContext)
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "papagaio_tts")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getSpeakingStatus" -> {
                result.success(getSpeakingStatus())
            }

            "getVoices" -> {
                result.success(getVoices())
            }

            "getAvailableLanguages" -> {
                val filterLanguages = call.arguments as List<String>
                result.success(getAvailableLanguages(filterLanguages))
            }

            "getLanguage" -> {
                result.success(getLanguage())
            }

            "getVoice" -> {
                result.success(getVoice())
            }

            "getRate" -> {
                result.success(getRate())
            }

            "getPitch" -> {
                result.success(getPitch())
            }

            "getVolume" -> {
                result.success(getVolume())
            }

            "setLanguage" -> {
                val language = call.arguments as? List<String>
                println("language $language")
                if (language != null) {
                    result.success(setLanguage(language))
                } else {
                    result.success(false)
                }
            }

            "setVoice" -> {
                val voice = call.arguments.toString()
                if (voice != null) {
                    result.success(setVoice(voice))
                } else {
                    result.success(false)
                }
            }

            "setRate" -> {
                val rate = call.arguments.toString().toFloat()
                if (rate != null) {
                    result.success(setRate(rate))
                } else {
                    result.success(false)
                }
            }

            "setPitch" -> {
                val pitch = call.arguments.toString().toFloat()
                if (pitch != null) {
                    result.success(setPitch(pitch))
                } else {
                    result.success(false)
                }
            }

            "setVolume" -> {
                val volume = call.arguments.toString().toFloat()
                if (volume != null) {
                    result.success(setVolume(volume))
                } else {
                    result.success(false)
                }
            }

            "speak" -> {
                val text = call.arguments.toString().toString()
                result.success(speak(text))
            }

            "stop" -> {
                result.success(stop())
            }

            "shutdown" -> {
                shutdown()
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun speak(text: String): Boolean {
        return tts.speak(text, volume)
    }

    private fun getSpeakingStatus(): Boolean {
        return tts.getSpeakingStatus()
    }

    private fun getAvailableLanguages(filterLanguages: List<String>): List<List<String>> {
        return tts.getAvailableLanguages(filterLanguages)
    }

    private fun getVoices(): List<String> {
        return tts.getVoices()
    }

    private fun getLanguage(): String {
        return tts.getLanguage()
    }

    private fun getVoice(): String {
        return tts.getVoice()
    }

    private fun getRate(): Float {
        return this.rate
    }

    private fun getPitch(): Float {
        return this.pitch
    }

    private fun getVolume(): Float {
        return this.volume
    }

    private fun setVoice(name: String): Boolean {
        return tts.setVoice(name)
    }

    private fun setLanguage(language: List<String>): Boolean {
        return tts.setLanguage(language[0], language[1])
    }

    private fun setRate(rate: Float): Boolean {
        val result = tts.setRate(rate)
        if (result) {
            this.rate = rate
            return true
        }
        return false
    }

    private fun setPitch(pitch: Float): Boolean {
        val result = tts.setPitch(pitch)
        if (result) {
            this.pitch = pitch
            return true
        }
        return false
    }

    private fun setVolume(volume: Float): Boolean {
        this.volume = volume
        return true
    }

    private fun stop(): Boolean {
        return tts.stop()
    }

    private fun shutdown() {
        tts.shutdown()
    }
}
